x-pack/plugins/secret_service/server/lib/secret_service.ts

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import crypto from 'crypto';
+import { SavedObjectsClient } from 'src/legacy/server/saved_objects';
+// @ts-ignore
+import { AuditLogger } from '../../../server/lib/audit_logger';
+import { buildCrypt } from './crypt_keeper';
+
+type SSEvents = 'secret_object_created' | 'secret_object_accessed';
+
+export class SecretService {
+  public readonly hideAttribute: (deets: any, secretKey: string) => any;
+  public readonly unhideAttribute: (deets: any) => any;
+  private readonly type: string;
+  private readonly auditor?: AuditLogger;
+
+  constructor(
+    savedObjectsClient: SavedObjectsClient,
+    type: string,
+    key?: Buffer,
+    auditor?: AuditLogger
+  ) {
+    this.type = type;
+    this.auditor = auditor;
+    key = key || crypto.randomBytes(128);
```

Comment:
nit: if the key is required and always provided, I'd recommend throwing an Error explicitly here if we don't see a key.

Reply:
This makes a lot of sense, it would keep us from having odd behavior in case the calling code accidentally provides an `undefined` value for some reason.